src/main/java/main/movie/dao/impl/UserDaoImpl.java

```diff
@@ -0,0 +1,73 @@
+package main.movie.dao.impl;
+
+import java.util.List;
+import java.util.Optional;
+import javax.persistence.criteria.CriteriaBuilder;
+import javax.persistence.criteria.CriteriaQuery;
+import javax.persistence.criteria.Predicate;
+import javax.persistence.criteria.Root;
+import main.movie.dao.UserDao;
+import main.movie.exceptions.DataProcessingException;
+import main.movie.lib.Dao;
+import main.movie.model.User;
+import main.movie.util.HibernateUtil;
+import org.hibernate.Session;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+
+@Dao
+public class UserDaoImpl implements UserDao {
+
+    @Override
+    public User create(User user) {
+        Transaction transaction = null;
+        Session session = null;
+        try {
+            session = HibernateUtil.getSessionFactory().openSession();
+            transaction = session.beginTransaction();
+            session.save(user);
+            transaction.commit();
+            return user;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new DataProcessingException("Can't insert user entity!" + user, e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public Optional<User> get(Long id) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            return Optional.ofNullable(session.get(User.class, id));
+        } catch (Exception e) {
+            throw new RuntimeException("Can't get user with id = " + id + " from DB!", e);
+        }
+    }
+
+    @Override
+    public List<User> getAll() {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            Query<User> getAllUsersQuery = session.createQuery(
+                    "from User", User.class);
+            return getAllUsersQuery.getResultList();
+        } catch (Exception e) {
+            throw new DataProcessingException("Can't get all users from DB!", e);
+        }
+    }
+
+    @Override
+    public Optional<User> getByEmail(String email) {
+        try (Session session = HibernateUtil.getSessionFactory().openSession()) {
+            CriteriaBuilder criteriaBuilder = session.getCriteriaBuilder();
+            CriteriaQuery<User> query = criteriaBuilder.createQuery(User.class);
+            Root<User> root = query.from(User.class);
+            Predicate emailPredicate = criteriaBuilder.equal(root.get("email"), email);
+            return session.createQuery(query.where(emailPredicate)).uniqueResultOptional();
+        }
+    }
+}
```

Comment:
we didn't have to implement getall and get methods here
just findByEmail (wrong naming) and add (not create)

Reply:
> we didn't have to implement getall and get methods here
> just findByEmail (wrong naming) and add (not create)

yes, let's go step by step